cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll

```diff
@@ -2817,6 +2855,68 @@ class TranslatedReuseExpr extends TranslatedNonConstantExpr {
   }
 }
 
+/**
+ * The IR translation of the destructor calls of the parent `TranslatedThrow`.
+ *
+ * This object does not itself generate the destructor calls. Instead, its
+ * children provide the actual calls, and this object ensures that we correctly
+ * exit with an `ExceptionEdge` after executing all the destructor calls.
+ */
+class TranslatedDestructorsAfterThrow extends TranslatedElement, TTranslatedDestructorsAfterThrow {
```

Comment:
I read this several times by now, but it's not clear to me what this is actually doing at the moment. My expectation would have been that this only comes in play when we conditionally construct things, because of my understanding where `thow`s can occur.

Reply:
It's actually not related to conditionally constructed things. Consider this example:
```cpp
void f() {
  throw 42;
}
```
the IR we construct for this is:
```
1| void f()
1|   Block 0
1|     v1_1(void)       = EnterFunction              :
1|     m1_2(unknown)    = AliasedDefinition          :
1|     m1_3(unknown)    = InitializeNonLocal         :
1|     m1_4(unknown)    = Chi                        : total:m1_2, partial:m1_3
3|     r3_1(glval<int>) = VariableAddress[#throw3:5] :
3|     r3_2(int)        = Constant[42]               :
3|     m3_3(int)        = Store[#throw3:5]           : &:r3_1, r3_2
3|     v3_4(void)       = ThrowValue                 : &:r3_1, m3_3
 |   Exception -> Block 1

1|   Block 1
1|     v1_5(void) = Unwind       :
1|     v1_6(void) = AliasedUse   : m1_3
1|     v1_7(void) = ExitFunction :
```
The important thing here being that there's an `Exception` edge to the `ExitFunction` block which marks the fact that we're exiting the function with an exception. It's important that this is an `Exception` edge because that means control-flow can be "caught" by a `catch` block. For example, if we have:
```cpp
int f()
{
    try {
        throw 42;
    } catch(...) {
        return 1;
    }
    return 0;
}
```
then the IR will look like:
```
1| int f()
1|   Block 0
1|     v1_1(void)       = EnterFunction              :
1|     m1_2(unknown)    = AliasedDefinition          :
1|     m1_3(unknown)    = InitializeNonLocal         :
1|     m1_4(unknown)    = Chi                        : total:m1_2, partial:m1_3
4|     r4_1(glval<int>) = VariableAddress[#throw4:9] :
4|     r4_2(int)        = Constant[42]               :
4|     m4_3(int)        = Store[#throw4:9]           : &:r4_1, r4_2
4|     v4_4(void)       = ThrowValue                 : &:r4_1, m4_3
 |   Exception -> Block 1

5|   Block 1
5|     v5_1(void)       = CatchAny                 :
6|     r6_1(glval<int>) = VariableAddress[#return] :
6|     r6_2(int)        = Constant[1]              :
6|     m6_3(int)        = Store[#return]           : &:r6_1, r6_2
1|     r1_5(glval<int>) = VariableAddress[#return] :
1|     v1_6(void)       = ReturnValue              : &:r1_5, m6_3
1|     v1_7(void)       = AliasedUse               : m1_3
1|     v1_8(void)       = ExitFunction             :
```
That is, control-flow now ensures that we flow from `Throw` to the `CatchAny` block. Now, consider what happens when we need to destruct something before we take the `Exception` edge out of Block 0:
```cpp
struct S { ~S(); };

int f()
{
    try {
        S s;
        throw 42;
    } catch(...) {
        return 1;
    }
    return 0;
}
```
In this case, we need to destruct `s` before we transfer control-flow to the `catch` block. However, we need to "remember" that we're in an exceptional context so that we actually produce this `Exception` edge once we're done with all the destructor calls. This commit ensures that we do exactly that. So after this commit the IR looks like:
```
3| int f()
3|   Block 0
3|     v3_1(void)           = EnterFunction                   :
3|     m3_2(unknown)        = AliasedDefinition               :
3|     m3_3(unknown)        = InitializeNonLocal              :
3|     m3_4(unknown)        = Chi                             : total:m3_2, partial:m3_3
6|     r6_1(glval<S>)       = VariableAddress[s]              :
6|     m6_2(S)              = Uninitialized[s]                : &:r6_1
7|     r7_1(glval<int>)     = VariableAddress[#throw7:9]      :
7|     r7_2(int)            = Constant[42]                    :
7|     m7_3(int)            = Store[#throw7:9]                : &:r7_1, r7_2
7|     v7_4(void)           = ThrowValue                      : &:r7_1, m7_3
8|     r8_1(glval<S>)       = VariableAddress[s]              :
8|     r8_2(glval<unknown>) = FunctionAddress[~S]             :
8|     v8_3(void)           = Call[~S]                        : func:r8_2, this:r8_1
8|     m8_4(unknown)        = ^CallSideEffect                 : ~m3_4
8|     m8_5(unknown)        = Chi                             : total:m3_4, partial:m8_4
8|     v8_6(void)           = ^IndirectReadSideEffect[-1]     : &:r8_1, m6_2
8|     m8_7(S)              = ^IndirectMayWriteSideEffect[-1] : &:r8_1
8|     m8_8(S)              = Chi                             : total:m6_2, partial:m8_7
 |   Exception -> Block 1

8|   Block 1
8|     v8_9(void)       = CatchAny                 :
9|     r9_1(glval<int>) = VariableAddress[#return] :
9|     r9_2(int)        = Constant[1]              :
9|     m9_3(int)        = Store[#return]           : &:r9_1, r9_2
3|     r3_5(glval<int>) = VariableAddress[#return] :
3|     v3_6(void)       = ReturnValue              : &:r3_5, m9_3
3|     v3_7(void)       = AliasedUse               : ~m8_5
3|     v3_8(void)       = ExitFunction             :
```

That's exactly what we're seeing in the subsequent commit that accepts the test cases. Specifically, see [this](https://github.com/github/codeql/pull/16125/commits/b6ddb97e40099df42c1c4ea4f954458f074b0a6d#diff-ee90324f6086fb05698de8216d15dc09893146e61f91ad400f36f4a1370020dcL15929-R15966) part of the diff. Prior to the fix we simply generated a `ReturnVoid` without any `Exception` edge. However, after the fix we correctly generate a `Exception` edge that flows to the block starting with `Unwind`.

Does that make sense?

Reply:
Yeah, that's correct. I should be able to pull this out and move it to a separate PR if you prefer. I was being a bit lazy here, I guess

Reply:
Just leave it here.